File: benchmarks/remediation/fixtures/python-sql-psycopg-format/tests/verify.py
"""Two invocation modes.

  python3 tests/verify.py --exploit | --behavior   Sandbox mode: checks the app.py in this tree.
  python3 tests/verify.py <original> <repaired>    Harness mode: compares the checked-in trees.
"""
import importlib.util
import json
import os
import sys
import types

ATTACK = "open' OR '1'='1"
LEGITIMATE = "open"
ROWS = [(11, "Printer jam")]

psycopg = types.ModuleType("psycopg")
psycopg.connect = lambda dsn: None
sys.modules.setdefault("psycopg", psycopg)


class RecordingCursor:
    def __init__(self):
        self.queries = []

    def execute(self, sql, params=None):
        self.queries.append((str(sql), params))
        return self

    def fetchall(self):
        return list(ROWS)


def load(path, name="app_under_test"):
    spec = importlib.util.spec_from_file_location(name, path)
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


def local_path():
    return os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "app.py")


def exploit():
    cursor = RecordingCursor()
    load(local_path()).find_tickets(cursor, ATTACK)
    sql, params = cursor.queries[0]
    if ATTACK in sql:
        sys.stderr.write("vulnerability present: untrusted input is interpolated into SQL text\n")
        sys.exit(1)
    assert params is not None and ATTACK in list(params)
    sys.exit(0)


def behavior():
    cursor = RecordingCursor()
    rows = load(local_path()).find_tickets(cursor, LEGITIMATE)
    sql, params = cursor.queries[0]
    assert "FROM tickets" in sql
    assert LEGITIMATE in sql or (params is not None and LEGITIMATE in list(params))
    assert rows == ROWS
    sys.exit(0)


def compare(original_path, repaired_path):
    original, repaired, legitimate = RecordingCursor(), RecordingCursor(), RecordingCursor()
    load(original_path, "original").find_tickets(original, ATTACK)
    load(repaired_path, "repaired").find_tickets(repaired, ATTACK)
    load(repaired_path, "repaired_legit").find_tickets(legitimate, LEGITIMATE)
    assert "OR '1'='1" in original.queries[0][0]
    assert repaired.queries[0] == ("SELECT id, subject FROM tickets WHERE status = %s", (ATTACK,))
    assert legitimate.queries[0][1] == (LEGITIMATE,)
    sys.stdout.write(json.dumps({"vulnerability_observed": True, "behavior_preserved": True}))


if __name__ == "__main__":
    mode = sys.argv[1] if len(sys.argv) > 1 else ""
    if mode == "--exploit":
        exploit()
    elif mode == "--behavior":
        behavior()
    else:
        compare(sys.argv[1], sys.argv[2])

File: benchmarks/remediation/fixtures/js-sql-exports-assignment/tests/verify.js
// Two invocation modes.
//   node tests/verify.js --exploit | --behavior   Sandbox mode: checks the app.js in this tree.
//   node tests/verify.js <original> <repaired>    Harness mode: compares the checked-in trees.
const assert = require('node:assert/strict');
const path = require('node:path');

const ATTACK = "x' OR '1'='1";
const LEGITIMATE = 'person@example.com';

// A connection that records what it was asked to run, so the test can tell an interpolated
// query from a parameterized one without a database.
function recorder() {
  const calls = [];
  return { calls, query: (text, values) => (calls.push({ text: String(text), values }), { rows: [] }) };
}

function run(module, email) {
  const db = recorder();
  module.findCustomer(db, email);
  return db.calls[0];
}

function loadLocal() {
  return require(path.resolve(__dirname, '..', 'app.js'));
}

function exploit() {
  const call = run(loadLocal(), ATTACK);
  if (call.text.includes(ATTACK)) {
    process.stderr.write('vulnerability present: untrusted input is interpolated into SQL text\n');
    process.exit(1);
  }
  assert.ok(Array.isArray(call.values) && call.values.includes(ATTACK));
  process.exit(0);
}

function behavior() {
  const call = run(loadLocal(), LEGITIMATE);
  assert.match(call.text, /FROM customers/i);
  assert.ok(call.text.includes(LEGITIMATE) || (Array.isArray(call.values) && call.values.includes(LEGITIMATE)));
  process.exit(0);
}

function compare(originalPath, repairedPath) {
  const original = run(require(originalPath), ATTACK);
  const repaired = run(require(repairedPath), ATTACK);
  const legitimate = run(require(repairedPath), LEGITIMATE);
  assert.ok(original.text.includes(ATTACK), 'the original interpolates the payload');
  assert.ok(!repaired.text.includes(ATTACK), 'the repair keeps the payload out of the SQL text');
  assert.deepEqual(repaired.values, [ATTACK]);
  assert.deepEqual(legitimate.values, [LEGITIMATE]);
  process.stdout.write(JSON.stringify({ vulnerability_observed: true, behavior_preserved: true }));
}

const mode = process.argv[2];
if (mode === '--exploit') exploit();
else if (mode === '--behavior') behavior();
else compare(process.argv[2], process.argv[3]);
